README.md
# React Native Radio Trimmer


## Installation

## package installation
```
yarn add react-native-radio-trimmer
```
or
```
npm install react-native-radio-trimmer
```
### install dependencies
you need to install additional libs in your project locally
```
yarn add react-native-svg@9.13.3
yarn add react-native-safe-area-view@0.14.8

```
### install pods in your project folder
```
cd ios && pod install && cd ../
```

## Working examples

### ticksColor
<p align=center>
  <img src="./gifs/colorOfTicks.gif">
</p> 

### ticksCount
<p align=center>
  <img src="./gifs/countTicks.gif">
</p> 

### dotColor
<p align=center>
  <img src="./gifs/dotColor.gif">
</p> 

### ticksCountHover
<p align=center>
  <img src="./gifs/countHoverTicks.gif">
</p> 

### ticksLength
<p align=center>
  <img src="./gifs/tickLength.gif">
</p> 

### ticksWidth
<p align=center>
  <img src="./gifs/tickWidth.gif">
</p> 

### dotIsShadow
<p align=center>
  <img src="./gifs/shadow1.gif">
</p> 

### pathColor
<p align=center>
  <img src="./gifs/pathColor.gif">
</p> 

## Usage

It is necessary to use onMount method, otherwise app will crash

```
import React, { Component } from 'react';
import { SafeAreaView, View, Text } from 'react-native';
import RadioTrimmer from 'react-native-radio-trimmer';


export class ExampleComponent extends Component {
  constructor(props) {
    super(props);
    this.state = {
      textValue: null,
    };
  }
  setValue = value => {
    this.setState({
      textValue: value,
    });
  };

  render() {
    return (
      <SafeAreaView>
        <Header />
        <RadioTrimmer
          boxSize={150}
          marginTop={50}
          onChangeValue={this.setValue}
          onMount={this.setValue}
        >
          <View>
            <Text>{this.state.textValue}</Text>
          </View>
        </RadioTrimmer>
      </SafeAreaView>
    );
  }
}

export default ExampleComponent;
```

## Few trimmers in one row:
```
   <View style={{ flexDirection: 'row' }}>
          <RadioTrimmer
            boxSize={150}
            marginTop={50}
            onChangeValue={this.setValue}
            onMount={this.setValue}
          >
            <View>
              <Text>{this.state.textValue}</Text>
            </View>
          </RadioTrimmer>
          <RadioTrimmer
            boxSize={150}
            marginTop={50}
            onChangeValue={this.setValue}
            onMount={this.setValue}
          >
            <View>
              <Text>{this.state.textValue}</Text>
            </View>
          </RadioTrimmer>

          <RadioTrimmer
            boxSize={150}
            marginTop={50}
            onChangeValue={this.setValue}
            onMount={this.setValue}
          >
            <View>
              <Text>{this.state.textValue}</Text>
            </View>
          </RadioTrimmer>
        </View>

```


## Configuration
### props:
| Property | Type | Default | Description |
|---------------|----------|-------------|----------------------------------------------------------------|
| accuracy | number | 10 | accuracy of result. 10 will result with 0.1, 0.5, etc |
| allowTap | boolean | true | when tapped on a trimmer it changes value to the tapped place|
| backgroundColor | string | transparent | background color of box with Radio Trimmer |
| boxSize | number | 200 | size of box containing the Radio Trimmer |
| dotColor | string | gray | color of radio dot |
| dotIsShadow | boolean | false | makes the dot background blurred|
| dotSize | number | 24 | width and height of the dot |
| innerDotColor | string | white | color of another dot inside the radio dot |
| innerDotSize | number | 10 | default inner dot size  |
| marginTop | number | 0 | margin that allows us to position the box fro top |
| maxValue | number | 108 | maximum scope value |
| minValue | number | 87.5 | minimal scope value |
| onChangeValue | function | null | function that is invoked on every radio text change. it return value of current text. Look at example |
| onMount | function | null | callback function that sets initial value on parent component |
| pathColor | string | lightgray | color of the path under the radio dot |
| pathIsShadow | boolean | false | makes the background of radio trimmer blurred |
| pathWidth | number | 10 | width of the path under the radio dot |
| step | number | 0.1 | value that will be added to each iteriationof calculating text value |
| textAfterNumber | string | MHz | the text that will be added to calculated text value |
| textBackgroundColor | string | transparent | the color under the inside calculated text |
| tickMargin | number | 0 | the distance between radio path and ticks |
| ticksActiveColor | string | black | color of active (hovered) tick  |
| ticksColor | string | gray | color of inactive (unhovered) tick |
| ticksCount | number | 25 | how many ticks will be generated |
| ticksCountHover | number | 5 | how many ticks will be hovered when the radio dot is near |
| ticksLength | number | 15 | the height of single tick |
| ticksWidth | number | 2 | the width of single tick  |

### onChangeValue, onMount:
```
  setValue = value => {
    this.setState({
      textValue: value,
    });
  };

<RadioTrimmer
          boxSize={150}
          marginTop={50}
          onChangeValue={this.setValue}
          onMount={this.setValue}
        >
```





CHANGELOG.md
### 0.4.0 (2020-03-10)

##### Breaking Changes

*  api changes/added swipe functionality ([2da913c1](https://github.com/Lkubok/react-native-radio-trimmer/commit/2da913c16209d3cfc83f0d35e2de77c13f0e88d6))

#### 0.2.2 (2020-03-10)

##### Documentation Changes

*  added generate-changelog package ([a6fbe3a2](https://github.com/Lkubok/react-native-radio-trimmer/commit/a6fbe3a2fcad4a85fa30ef641eef97471845b1c9))

##### Other Changes

*  row funcionality ([e23ae5e4](https://github.com/Lkubok/react-native-radio-trimmer/commit/e23ae5e4f6916ae85591275b2c1595fd6e264362))



renderTicks.js
/* eslint-disable react-native/no-inline-styles */
import React from "react";
import { View } from "react-native";
const renderTicks = ({
  ticksCount,
  boxCenterX,
  boxCenterY,
  dotSize,
  pureRadius,
  normalizedAngle,
  ticksLength,
  ticksWidth,
  ticksCountHover,
  ticksColor,
  ticksActiveColor,
  tickMargin
}) => {
  const boxCenter = {
    x: boxCenterX + dotSize / 2,
    y: boxCenterY + dotSize / 2
  };

  let ticksArr = [];

  const initPos = {
    top: boxCenter.x,
    left: boxCenter.y - ticksWidth / 2
  };

  for (let i = 0; i < ticksCount; i++) {
    let angleStep = 360 / ticksCount;
    let curAngle = angleStep * i;
    let curAngleDeg = curAngle * 1;
    if (curAngle > 180) {
      curAngle -= 360;
    }
    curAngle = curAngle * (Math.PI / 180);

    let normCurAngle =
      curAngleDeg + 90 - 360 * Math.floor((curAngleDeg + 90) / 360);

    let isActive = false;
    let min = normalizedAngle - (angleStep * ticksCountHover) / 2 + 90;
    let max = normalizedAngle + (angleStep * ticksCountHover) / 2 + 90;
    if (max > 360) {
      max -= 360;
      min -= 360;
    }
    if (
      (normCurAngle >= min && normCurAngle <= max) ||
      (min < 0 && normCurAngle > 360 + min)
    ) {
      isActive = true;
    }
    ticksArr.push(
      <View
        key={curAngle}
        style={{
          backgroundColor: isActive ? ticksActiveColor : ticksColor,
          transform: [{ rotate: `${angleStep * i}deg` }],
          position: "absolute",
          top:
            initPos.top -
            pureRadius * Math.cos(curAngle) -
            ticksLength * Math.cos(curAngle) -
            tickMargin * Math.cos(curAngle) +
            -ticksLength / 2,
          left:
            initPos.left +
            pureRadius * Math.sin(curAngle) +
            ticksLength * Math.sin(curAngle) +
            tickMargin * Math.sin(curAngle),

          height: ticksLength,
          width: ticksWidth
        }}
      />
    );
  }

  return ticksArr;
};

export default renderTicks;


createStyles.js
import { StyleSheet } from "react-native";

const createStyles = ({
  dotColor,
  dotSize,
  boxSize,
  backgroundColor,
  textBackgroundColor,
  innerDotColor,
  innerDotSize,
  marginTop,
  pathColor,
  dotIsShadow,
  pathIsShadow
}) => {
  return StyleSheet.create({
    radioDot: {
      backgroundColor: dotColor,
      width: dotSize,
      height: dotSize,
      borderRadius: dotSize / 2,
      position: "absolute",
      shadowColor: dotIsShadow ? dotColor : null,
      shadowOffset: dotIsShadow ? { width: 1, height: 1 } : null,
      shadowOpacity: dotIsShadow ? 0.8 : null,
      shadowRadius: dotIsShadow ? 10 : null
    },
    viewBox: {
      justifyContent: "center",
      alignItems: "center",
      position: "relative",
      marginTop: marginTop,
      backgroundColor: backgroundColor,
      shadowColor: pathIsShadow ? pathColor : null,
      shadowOffset: pathIsShadow ? { width: -10, height: 15 } : null,
      shadowOpacity: pathIsShadow ? 0.4 : null,
      shadowRadius: pathIsShadow ? 15 : null
    },
    radioBox: {
      width: boxSize,
      height: boxSize,
      backgroundColor: "transparent",
      position: "relative",
      borderRadius: boxSize / 2
    },
    changingTextBox: {
      justifyContent: "center",
      alignItems: "center",
      position: "absolute",
      borderRadius: boxSize / 2,
      backgroundColor: textBackgroundColor
    },
    whiteDot: {
      backgroundColor: innerDotColor,
      position: "relative",
      width: innerDotSize,
      height: innerDotSize,
      borderRadius: innerDotSize / 2,
      top: (dotSize - innerDotSize) / 2,
      left: (dotSize - innerDotSize) / 2
    },
    tickBox: {
      width: boxSize,
      height: boxSize,
      position: "relative",
      borderRadius: boxSize / 2
    },
    textInfo: {
      position: "absolute"
    }
  });
};

export default createStyles;


RadioTrimmer.js
/* eslint-disable react-native/no-inline-styles */
import React, { Component } from "react";
import { View, Text } from "react-native";
import { Svg, Circle } from "react-native-svg";
import Animated from "react-native-reanimated";
import { getInset } from "react-native-safe-area-view";
import {
  PanGestureHandler,
  TapGestureHandler,
  State
} from "react-native-gesture-handler";
import PropTypes from "prop-types";

import renderTicks from "./renderTicks";
import createStyles from "./createStyles";

const { call, cond, eq, event, Value } = Animated;
export class RadioTrimmer extends Component {
  constructor(props) {
    super(props);
    const { boxSize, dotSize, minValue } = props;
    this.state = {
      angle: null,
      boxCenterX: boxSize / 2 - dotSize / 2,
      boxCenterY: boxSize / 2 - dotSize / 2,
      boxMountPosX: null,
      boxMountPosY: null,
      boxSize,
      centerTestX: null,
      centerTestY: null,
      calculatedText: minValue,
      dotOffsetX: boxSize / 2 - dotSize / 2,
      heightFromTopOfScreen: getInset("top"),
      dotOffsetY: 0,
      dotSize,
      normalizedAngle: null,
      pureRadius: boxSize / 2,
      radius: boxSize / 2 - dotSize / 2,
      ticksCount: this.props.ticksCount,
      moveInterval: null
    };
    this.absoluteX = new Value(0);
    this.absoluteY = new Value(0);
    this.gestureState = new Value(-1);
    this.onGestureEvent = event([
      {
        nativeEvent: {
          absoluteX: this.absoluteX,
          absoluteY: this.absoluteY,
          state: this.gestureState
        }
      }
    ]);
  }
  componentDidMount() {
    this.props.onMount(this.props.minValue);
  }
  handleLayoutChange = () => {
    this.RadioTrimmer.measure((px, py) => {
      this.setState({
        boxMountPosY: py
      });
    });
  };
  handleLayoutChangeX = () => {
    this.RadioTrimmer.measure((px, py, w, h, pax, pay) => {
      this.setState({
        centerTestX: pax,
        centerTestY: pay
      });
    });
  };
  handleTap = e => {
    const { absoluteX, absoluteY } = e.nativeEvent;
    this.moving([absoluteX, absoluteY]);
  };
  moving = ([curX, curY]) => {
    const { heightFromTopOfScreen, radius } = this.state;
    const { onChangeValue } = this.props;
    const boxCenter = {
      x: this.state.centerTestX,
      y: this.state.centerTestY + heightFromTopOfScreen
    };
    const pointerCurPos = {
      x: curX,
      y: curY
    };
    const calcX = Math.cos(this.mesAngle(boxCenter, pointerCurPos)) * radius;
    const calcY = Math.sin(this.mesAngle(boxCenter, pointerCurPos)) * radius;
    this.setState(prev => {
      return {
        dotOffsetX: prev.boxCenterX + calcX,
        dotOffsetY: prev.boxCenterY + calcY,
        angle: this.mesAngleDeg(boxCenter, pointerCurPos),
        normalizedAngle:
          this.mesAngleDeg(boxCenter, pointerCurPos) +
          90 -
          360 *
            Math.floor((this.mesAngleDeg(boxCenter, pointerCurPos) + 90) / 360)
      };
    });
    this.calculateRadioValue();
    onChangeValue ? this.props.onChangeValue(this.state.calculatedText) : null;
  };
  mesAngle = (p1, p2) => {
    return Math.atan2(p2.y - p1.y, p2.x - p1.x);
  };
  mesAngleDeg = (p1, p2) => {
    return (Math.atan2(p2.y - p1.y, p2.x - p1.x) * 180) / Math.PI;
  };
  calculateRadioValue = () => {
    const { minValue, maxValue, step, accuracy } = this.props;
    const { radius, angle, normalizedAngle } = this.state;
    const circuitLength = Math.PI * 2 * radius;
    const routeDistance = maxValue - minValue;
    const countOfSteps = routeDistance / step;
    const oneStepLength = circuitLength / countOfSteps;
    const lengthOnArc = (normalizedAngle / 360) * circuitLength;
    let ValueToReturn = angle
      ? minValue + Math.ceil(lengthOnArc / oneStepLength) * step
      : minValue;
    this.setState({
      calculatedText: Math.floor(ValueToReturn * accuracy) / accuracy
    });
  };
  ticksRender = () => {
    return renderTicks({ ...this.state, ...this.props });
  };
  render() {
    const {
      backgroundColor,
      boxSize,
      dotColor,
      dotIsShadow,
      dotSize,
      innerDotColor,
      innerDotSize,
      marginTop,
      pathColor,
      pathIsShadow,
      pathWidth,
      textBackgroundColor,
      children
    } = this.props;
    const {
      dotOffsetX,
      dotOffsetY,
      calculatedText,
      boxCenterY,
      boxCenterX
    } = this.state;
    const styles = createStyles({
      backgroundColor,
      boxSize,
      dotColor,
      dotIsShadow,
      dotSize,
      innerDotColor,
      innerDotSize,
      marginTop,
      pathColor,
      pathIsShadow,
      pathWidth,
      textBackgroundColor
    });

    return (
      <TapGestureHandler
        onHandlerStateChange={this.props.allowTap ? this.handleTap : null}
        maxDurationMs={400}
      >
        <View style={styles.viewBox} onLayout={this.handleLayoutChange}>
          <View
            style={styles.radioBox}
            onLayout={e => {
              this.setState({
                boxMountPosX: e.nativeEvent.layout.x
              });
            }}
          >
            <View style={styles.changingTextBox}>
              <View style={styles.textInfo}>
                {children ? children : <Text>{calculatedText}</Text>}
              </View>
              <View style={styles.tickBox}>{this.ticksRender()}</View>
              <Svg
                blurRadius={20}
                height={boxSize}
                style={{ position: "absolute" }}
                width={boxSize}
              >
                <Circle
                  cx={boxSize / 2}
                  cy={boxSize / 2}
                  fill="transparent"
                  r={boxSize / 2 - dotSize / 2}
                  stroke={pathColor}
                  strokeWidth={pathWidth}
                />
              </Svg>
              <View
                style={{
                  position: "absolute",
                  top: boxCenterY + dotSize / 2,
                  left: boxCenterX + dotSize / 2
                }}
                ref={view => {
                  this.RadioTrimmer = view;
                }}
                onLayout={this.handleLayoutChangeX}
              />
            </View>
            <Animated.Code>
              {() =>
                cond(
                  eq(this.gestureState, State.ACTIVE),
                  call([this.absoluteX, this.absoluteY], this.moving)
                )
              }
            </Animated.Code>

            <PanGestureHandler
              maxpointerCurPoss={1}
              minDist={10}
              onGestureEvent={this.onGestureEvent}
              onHandlerStateChange={this.onGestureEvent}
            >
              <Animated.View
                style={[
                  styles.radioDot,
                  {
                    top: dotOffsetY,
                    left: dotOffsetX
                  }
                ]}
              >
                <View style={styles.whiteDot} />
              </Animated.View>
            </PanGestureHandler>
          </View>
        </View>
      </TapGestureHandler>
    );
  }
}

RadioTrimmer.propTypes = {
  accuracy: PropTypes.number,
  allowTap: PropTypes.bool,
  backgroundColor: PropTypes.string,
  boxSize: PropTypes.number,
  children: PropTypes.Component,
  dotColor: PropTypes.string,
  dotIsShadow: PropTypes.bool,
  dotSize: PropTypes.number,
  innerDotColor: PropTypes.string,
  innerDotSize: PropTypes.number,
  marginTop: PropTypes.number,
  maxValue: PropTypes.number,
  minValue: PropTypes.number,
  onChangeValue: PropTypes.func,
  onMount: PropTypes.func,
  pathColor: PropTypes.string,
  pathIsShadow: PropTypes.bool,
  pathWidth: PropTypes.number,
  step: PropTypes.number,
  textAfterNumber: PropTypes.string,
  textBackgroundColor: PropTypes.string,
  tickMargin: PropTypes.number,
  ticksActiveColor: PropTypes.string,
  ticksColor: PropTypes.string,
  ticksCount: PropTypes.number,
  ticksCountHover: PropTypes.number,
  ticksLength: PropTypes.number,
  ticksWidth: PropTypes.number
};

RadioTrimmer.defaultProps = {
  accuracy: 10,
  allowTap: true,
  backgroundColor: "transparent",
  boxSize: 200,
  dotColor: "gray",
  dotIsShadow: false,
  dotSize: 24,
  innerDotColor: "white",
  innerDotSize: 10,
  marginTop: 0,
  maxValue: 108,
  minValue: 87.5,
  onChangeValue: null,
  pathColor: "lightgray",
  pathIsShadow: false,
  pathWidth: 10,
  step: 0.1,
  textAfterNumber: "MHz",
  textBackgroundColor: "transparent",
  tickMargin: 0,
  ticksActiveColor: "black",
  ticksColor: "gray",
  ticksCount: 25,
  ticksCountHover: 5,
  ticksLength: 15,
  ticksWidth: 2
};

export default RadioTrimmer;
